### PR TITLE
Refactor api calls

### PR DESCRIPTION
**Pull request by IgorZiegler**
Refactor few api calls

### CODE REVIEW

File: js/src/chrome/settings/chrome.settings.cartocss.js [hunk 1]
```diff
@@ -272,10 +272,10 @@ Wu.Chrome.SettingsContent.Cartocss = Wu.Chrome.SettingsContent.extend({
 			file_id: file_id,
 			return_model : true,
 			layerUuid : layer.getUuid()
-		}
+		};
 
 		// create layer on server
-		Wu.post('/api/db/createLayer', JSON.stringify(layerJSON), function (err, newLayerJSON) {
+		app.api.createLayer(layerJSON, function (err, newLayerJSON) {
```
Comment (knutole) [inline js/src/chrome/settings/chrome.settings.cartocss.js]: we should check for errors on all callbacks

File: js/src/chrome/settings/chrome.settings.extras.js [hunk 2]
```diff
@@ -439,7 +439,7 @@ Wu.Chrome.SettingsContent.Extras = Wu.Chrome.SettingsContent.extend({
 		console.log('layerJSON', layerJSON);
 
 		// create layer on server
-		Wu.post('/api/db/createLayer', JSON.stringify(layerJSON), function (err, newLayerJSON) {
+		app.api.createLayer(layerJSON, function (err, newLayerJSON) {
```
Comment (knutole) [inline js/src/chrome/settings/chrome.settings.extras.js]: check for err

File: js/src/chrome/settings/chrome.settings.filters.js [hunk 3]
```diff
@@ -191,10 +191,10 @@ Wu.Chrome.SettingsContent.Filters = Wu.Chrome.SettingsContent.extend({
 			file_id: file_id,
 			return_model : true,
 			layerUuid : layer.getUuid()
-		}
+		};
 
 		// create layer on server
-		Wu.post('/api/db/createLayer', JSON.stringify(layerJSON), function (err, newLayerJSON) {
+		app.api.createLayer(layerJSON, function (err, newLayerJSON) {
```
Comment (knutole) [inline js/src/chrome/settings/chrome.settings.filters.js]: check for err

File: js/src/chrome/settings/styler.js [hunk 4]
```diff
@@ -1147,10 +1147,10 @@ Wu.Styler = Wu.Class.extend({
 			file_id: file_id,
 			return_model : true,
 			layerUuid : layer.getUuid()
-		}
+		};
 
 		// create layer on server
-		Wu.post('/api/db/createLayer', JSON.stringify(layerJSON), function (err, newLayerJSON) {
+		app.api.createLayer(layerJSON, function (err, newLayerJSON) {
```
Comment (knutole) [inline js/src/chrome/settings/styler.js]: check for err

File: js/src/chrome/users/chrome.users.js [hunk 5]
```diff
@@ -279,14 +279,14 @@ Wu.Chrome.Users = Wu.Chrome.extend({
...
-		var options = JSON.stringify({
+		var options = {
 			access : access
-		});
+		};
 
 		// create share link
-		Wu.post('/api/invite/link', options, function (a, b) {
+		app.api.inviteLink(options, function (a, b) {
 			this._shareLinkInput.value = b;
```
Comment (knutole) [inline js/src/chrome/users/chrome.users.js]: callback returns (err, result). should be renamed to such.

File: js/src/models/model.layers.js [hunk 6]
```diff
@@ -553,8 +553,17 @@ Wu.Model.Layer = Wu.Model.extend({
 	},
 
 	_save : function (json) {
-		var string  = JSON.stringify(json);
-		Wu.save('/api/layer/update', string);
+		app.api.updateLayer(json, function (err, result) {
+			if (err) console.error('err', err);
```
Comment (knutole) [inline js/src/models/model.layers.js]: better error handling here - give user feedback if something went wrong.

`app.feedback.setError()` (don't quite remember correct syntax, but look in `/js/src/panes/pane.feedback.js`

File: js/src/models/model.projects.js [hunk 7]
```diff
@@ -485,8 +485,18 @@ Wu.Project = Wu.Class.extend({
 			project_id : this.getUuid(),
 			layer_id : lids
 		};
-		var string = JSON.stringify(json);
-		Wu.save('/api/layers/delete', string); 
+
+		app.api.deleteLayer(json, function (err, result) {
+			if (err) console.error('err', err);
```
Comment (knutole) [inline js/src/models/model.projects.js]: should have better error handling, with user feedback

**Comment by knutole** [comment]
Should also remove console.error on line model.layers.js:565 before merge

**Comment by knutole** [comment]
@IgorZiegler You can grep  in `/docks/modules/wu/public/` to find more API calls that need to be updated.
- `grep -iR . -e 'Wu.save'`
- `grep -iR . -e 'Wu.post'`
- `grep -iR . -e 'Wu.send'`

File: js/src/models/model.users.js [hunk 8]
```diff
@@ -60,21 +60,16 @@ Wu.User = Wu.Class.extend({
 
 		var options = {
 			contact : user.getUuid()
-		}
-
-		Wu.send('/api/user/requestContact', options, function (a, b) {
+		};
 
+		app.api.requestContact(options, function (a, b) {
```
Comment (knutole) [inline js/src/models/model.users.js]: a, b should be err, result

File: js/src/models/model.users.js [hunk 9]
```diff
@@ -88,10 +83,10 @@ Wu.User = Wu.Class.extend({
 			edit : options.edit,
 			read : options.read,
 			user : userUuid
-		}
+		};
 
 		// send to server
-		Wu.send('/api/user/inviteToProjects', invites, function (a, response) {
+		app.api.inviteToProjects(invites, function (a, response) {
```
Comment (knutole) [inline js/src/models/model.users.js]: args should be err, result

File: js/src/panes/pane.share.js [hunk 10]
```diff
@@ -181,9 +181,9 @@ Wu.Share = Wu.Pane.extend({
 		app.setHash(function (ctx, hash) {
 
 			// get snapshot from server
-			Wu.send('/api/util/snapshot', hash, function (a, b) {
+			app.api.utilSnapshot(hash, function (a, b) {
```
Comment (knutole) [inline js/src/panes/pane.share.js]: args should be err, result

File: js/src/models/model.projects.js [hunk 11]
```diff
@@ -485,8 +485,23 @@ Wu.Project = Wu.Class.extend({
 			project_id : this.getUuid(),
 			layer_id : lids
 		};
-		var string = JSON.stringify(json);
-		Wu.save('/api/layers/delete', string); 
+
+		app.api.deleteLayer(json, function (err, result) {
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in removeMapboxAccount',
```
Comment (knutole) [inline js/src/models/model.projects.js]: This error message seems a bit too specific?? ;)

File: js/src/chrome/settings/chrome.settings.cartocss.js [hunk 12]
```diff
@@ -272,11 +272,16 @@ Wu.Chrome.SettingsContent.Cartocss = Wu.Chrome.SettingsContent.extend({
...
-
+		app.api.createLayer(layerJSON, function (err, newLayerJSON) {
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in _updateLayer',
+					description : err
+				});
+			}
 			// new layer
 			var newLayerStyle = Wu.parse(newLayerJSON);
```
Comment (knutole) [inline js/src/chrome/settings/chrome.settings.cartocss.js:288]: done() below in line 290 should probably return `done(err)` if there is an error....

File: js/src/chrome/settings/chrome.settings.extras.js [hunk 13]
```diff
@@ -439,8 +439,13 @@ Wu.Chrome.SettingsContent.Extras = Wu.Chrome.SettingsContent.extend({
...
-
+		app.api.createLayer(layerJSON, function (err, newLayerJSON) {
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in _updateLayer',
+					description : err
+				});
+			}
 			// new layer
 			var newLayerStyle = Wu.parse(newLayerJSON);
```
Comment (knutole) [inline js/src/chrome/settings/chrome.settings.extras.js:452]: done() below in line 454 should probably return done(err) if there is an error....

File: js/src/chrome/users/chrome.users.js [hunk 14]
```diff
@@ -279,14 +279,21 @@ Wu.Chrome.Users = Wu.Chrome.extend({
...
+		};
 
 		// create share link
-		Wu.post('/api/invite/link', options, function (a, b) {
-			this._shareLinkInput.value = b;
-		}.bind(this), this);
+		app.api.inviteLink(options, function (err, response) {
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in _createShareableInvite',
```
Comment (knutole) [inline js/src/chrome/users/chrome.users.js]: error message too specific. Remember, these messages are for the user.

`Something went wrong`, is enough for user feedback. (Same for all other places where similar message.)

File: js/src/controls/control.cartocss.js [hunk 15]
```diff
@@ -1178,10 +1178,16 @@ L.Control.Cartocss = Wu.Control.extend({
...
 			layerUuid : layer.getUuid()
-		}
+		};
 
-		Wu.post('/api/db/createLayer', JSON.stringify(layerJSON), function (err, newLayerJSON) {
+		app.api.createLayer(layerJSON, function (err, newLayerJSON) {
 
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in _updateLayer',
```
Comment (knutole) [inline js/src/controls/control.cartocss.js]: too specific error message

File: js/src/core/controller.js [hunk 16]
```diff
@@ -69,12 +69,19 @@ Wu.Controller = Wu.Class.extend({
...
-		}
+		};
 
 		// get a saved setup - which layers are active, position, 
-		Wu.post('/api/project/hash/get', JSON.stringify(json), function (ctx, reply) {
+		app.api.getHash(json, function (err, response) {
+
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in _loadState',
```
Comment (knutole) [inline js/src/core/controller.js]: err msg too specific

File: js/src/core/controller.js [hunk 17]
```diff
@@ -115,12 +122,20 @@ Wu.Controller = Wu.Class.extend({
...
-		}
+		};
 
 		// save hash to server
-		Wu.post('/api/project/hash/set', JSON.stringify(json), function (a, b) {
+		app.api.hashSet(json, function (err, response) {
+
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in _saveState',
```
Comment (knutole) [inline js/src/core/controller.js]: err msg too specific

File: js/src/models/model.files.js [hunk 18]
```diff
@@ -281,10 +281,16 @@ Wu.Model.File = Wu.Model.extend({
...
 	
 	_getLayers : function (callback) {
-
 		// get layers connected to dataset
 		var options = this._getLayerData();
-		Wu.post('/api/file/getLayers', JSON.stringify(options), function (err, response) {
+		app.api.fileGetLayers(options, function (err, response) {
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in _getLayers',
```
Comment (knutole) [inline js/src/models/model.files.js]: err msg too speciifc

File: js/src/models/model.files.js [hunk 19]
```diff
@@ -339,13 +345,19 @@ Wu.Model.File = Wu.Model.extend({
...
 
 		// set download id for feedback
 		this._downloadingID = Wu.Util.createRandom(5);
 
 		// post download request to server
-		Wu.post('/api/file/downloadDataset', JSON.stringify(options), function (err, response) {
+		app.api.downloadDataset(options, function (err, response) {
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in _downloadDataset',
```
Comment (knutole) [inline js/src/models/model.files.js]: err msg too specific

File: js/src/models/model.files.js [hunk 20]
```diff
@@ -582,10 +594,17 @@ Wu.Model.File = Wu.Model.extend({
...
 			"cutColor" : cutColor
-		}
+		};
 
 		// create postgis layer
-		Wu.post('/api/db/createLayer', JSON.stringify(layerJSON), function (err, layerJSON) {
+		app.api.createLayer(layerJSON, function (err, layerJSON) {
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in _requestDefaultRasterLayer',
```
Comment (knutole) [inline js/src/models/model.files.js]: err msg too specific

File: js/src/models/model.files.js [hunk 21]
```diff
@@ -676,10 +695,16 @@ Wu.Model.File = Wu.Model.extend({
...
 			"projectUuid" : project.getUuid()
-		}
+		};
 
 		// create postgis layer
-		Wu.post('/api/db/createLayer', JSON.stringify(layerJSON), function (err, layerJSON) {
+		app.api.createLayer(layerJSON, function (err, layerJSON) {
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in _requestDefaultVectorLayer',
```
Comment (knutole) [inline js/src/models/model.files.js]: err msg too specific

File: js/src/models/model.layers.js [hunk 22]
```diff
@@ -553,8 +552,23 @@ Wu.Model.Layer = Wu.Model.extend({
...
+					title : 'Something went wrong in _save',
+					description : err
+				});
+			}
+
+			var result = Wu.parse(result);
+
+			if (result.error) {
+				return app.feedback.setError({
+					title : 'Something went wrong in _save',
```
Comment (knutole) [inline js/src/models/model.layers.js]: err msg too specific

File: js/src/models/model.layers.js [hunk 23]
```diff
@@ -553,8 +552,23 @@ Wu.Model.Layer = Wu.Model.extend({
 	},
 
 	_save : function (json) {
-		var string  = JSON.stringify(json);
-		Wu.save('/api/layer/update', string);
+		app.api.updateLayer(json, function (err, result) {
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in _save',
```
Comment (knutole) [inline js/src/models/model.layers.js]: err msg too specific

File: js/src/models/model.layers.js [hunk 24]
```diff
@@ -860,13 +874,18 @@ Wu.PostGISLayer = Wu.Model.Layer.extend({
...
 
 		// set download id for feedback
 		this._downloadingID = Wu.Util.createRandom(5);
 
-		Wu.post('/api/layer/downloadDataset', JSON.stringify(options), function (err, resp) {
-
+		app.api.downloadLayerDataset(options, function (err, resp) {
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in downloadLayer',
```
Comment (knutole) [inline js/src/models/model.layers.js]: err msg too specific

File: js/src/models/model.projects.js [hunk 25]
```diff
@@ -498,7 +513,12 @@ Wu.Project = Wu.Class.extend({
 		};
 
 		app.api.deleteLayer('/api/layers/delete', JSON.stringify(options), function (err, response) {
-			if (err) return console.error('layer delete err:', err);
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in deleteLayer',
```
Comment (knutole) [inline js/src/models/model.projects.js]: err msg too specific

File: js/src/models/model.projects.js [hunk 26]
```diff
@@ -485,8 +485,23 @@ Wu.Project = Wu.Class.extend({
...
+				return app.feedback.setError({
+					title : 'Something went wrong in removeMapboxAccount',
+					description : err
+				});
+			}
+
+			var result = Wu.parse(result);
+
+			if (result.error) {
+				console.error('something went worng', result);
```
Comment (knutole) [inline js/src/models/model.projects.js]: should give user error feedback here also

File: js/src/models/model.users.js [hunk 27]
```diff
@@ -88,11 +88,16 @@ Wu.User = Wu.Class.extend({
...
-		}
+		};
 
 		// send to server
-		Wu.send('/api/user/inviteToProjects', invites, function (a, response) {
-
+		app.api.inviteToProjects(invites, function (err, response) {
+			if (err) {
+				return app.feedback.setError({
+					title : 'Something went wrong in inviteToProjects',
```
Comment (knutole) [inline js/src/models/model.users.js]: err msg too specific

File: js/src/panes/pane.share.js [hunk 28]
```diff
@@ -181,9 +181,16 @@ Wu.Share = Wu.Pane.extend({
 		app.setHash(function (ctx, hash) {
 
 			// get snapshot from server
-			Wu.send('/api/util/snapshot', hash, function (a, b) {
-				this._createdImage(a, b);
-			}.bind(this), this);
+			app.api.utilSnapshot(hash, function (err, response) {
+				if (err) {
+					return app.feedback.setError({
+						title : 'Something went wrong in _shareImage',
```
Comment (knutole) [inline js/src/panes/pane.share.js]: err msg too specific

File: js/src/chrome/settings/chrome.settings.cartocss.js [hunk 29]
```diff
@@ -272,11 +272,17 @@ Wu.Chrome.SettingsContent.Cartocss = Wu.Chrome.SettingsContent.extend({
...
 		// create layer on server
-		Wu.post('/api/db/createLayer', JSON.stringify(layerJSON), function (err, newLayerJSON) {
-
+		app.api.createLayer(layerJSON, function (err, newLayerJSON) {
+			if (err) {
+				app.feedback.setError({
+					title : 'Something went wrong',
+					description : err
+				});
+				return done(err);
```
Comment (knutole) [inline js/src/chrome/settings/chrome.settings.cartocss.js]: should make sure `done` exists, so `return done && done(err)`

**Comment by knutole** [comment]
very nice work, igor! :) are you ready for merge on this one?

**Comment by IgorZiegler** [comment]
@knutole updated. I think yep, ready to merge

**Pull request merged**